ai/build.gradle

```diff
@@ -3,4 +3,7 @@ description = 'TripleA ai implementations'
 dependencies {
     implementation project(":game-core")
     implementation project(":java-extras")
+    implementation project(":swing-lib")
```

Comment:
The swing dependencies in AI would be great to break. It would very much allow for an easier path to JavaFX and perhaps be better general decoupling of AI with UI.

Can we convert the dependencies to be data dependencies instead of actual UI components? What kinds of things do we need UI components for in this AI?

Reply:
It is the debug menu and dialogs.  Maybe I should create a "ai-swing-debug" package that is dependent on "swing-lib".  Then "ai" can depend on "ai-swing-debug" but use the new package for all of the debug ui creation stuff.

Reply:
That might just be punting the problem. Perhaps the game can provide an API that AI can use to send debug data and let the game handle the UI elements. Then it'll be up to game-core to bind data to UI.

Reply:
"swing-lib" brings in `SwingAction` which is used to handle `JMenuItem` actions.  So, if you don't want "swing-lib", then the ai can't build `JMenuItem`s.  Instead, I guess it will have to send back a map of `TripleaUiAction`s that have the name of the item, any child `TripleaUiAction`s, and whether it is a checkbox, radio button, or regular item.

The other 5 interactions with swing code that the debug uses is
```
TripleAFrame.getTerritoryDetails().addAdditionalTerritoryDetailsFunction(territoryDetailMethod);    TripleAFrame.getTerritoryDetails().removeAdditionalTerritoryDetailsFunction(territoryDetailMethod);
TripleAFrame.getMapPanel().setTerritoryOverlayForTile(territory, color, 100);
TripleAFrame.getMapPanel().clearTerritoryOverlay(territory);
TripleAFrame.getMapPanel().repaint();
```
(see `TerritoryDebugUiAction`)

I was also planning on having `LanchesterDebugUiAction` show the calculated `attributionFactor` or maybe hooking into the `BattleCalculator` somehow.  This isn't done yet.

Reply:
Just a few quick comments:
- `Map<Integer, Integer> = TreeMap<>`, List<Tuple<Integer, Integer>>`, `List<Point>` are all effectively the same data structure, the latter is the most descriptive though. To avoid javascript'esque typing, we can create interfaces that describe the data being passed. IE:
```
interface DebugMenuItem() {
   String getTitle();
   void renderContent(Consumer<Supplier> sendDataToMe);
}
```

We can probably do very similar to hide the 'mapPanel' and 'territory' details as well. EG:

```
Consumer<TerritoryName>  territoryClearFunction =  TripleAFrame.getMapPanel()::clearTerritoryOverlay
```
Is that workable at all? I think there just needs to be an inversion, java 8 functional interfaces potentially being very useful here to bind specific method calls to functions needed by AI. One example where this was used was with LobbyWatcher, instead of having a HeadlessLobbyWatcher and Headed, the only difference needing to be how errors were handled, the two were unified and accepted a `Consumer<String> errorReporter` that was in one case bound to a UI window and in the other was just a Logger call.

Reply:
I'm not exactly sure what the first comment is about.

But the `DebugMenuItem` interface won't cut it.  You'd need a class like:
```
@Value
class UiMenuItem {
  enum MenuType {
    CHECKBOX,
    RADIOBUTTON,
    DEFAULT,
    MENU,
  }
  String title;
  MenuType menuType;
  List<UiMenuItem> children;
  Consumer<DebugMenuActions> performAction;
}
```
Then `DebugMenuActions` would look like:
```
interface DebugMenuActions {
  Consumer<Map<Territory, String>> addAdditionalTerritoryDetails;
  Consumer<Map<Territory, String>> removeAdditionalTerritoryDetails;
  Consumer<TerritoryOverlayDetails> addTerritoryOverlay;
  @Value
  class TerritoryOverlayDetails {
    Territory territory;
    Color color; // Color is awt so that might need to change to some other abstraction
    int alpha;
  };
  Consumer<Territory> clearTerritoryOverlay;
  Runnable repaintMap;
}
```
The `DebugMenuActions`, though, will slowly get bigger as other debug requirements are added.

Reply:
And that api doesn't come close to encompassing what the ProAi needs for its debug menu and dialog.  It needs to not only create menu items but a dialog with buttons, text areas, checkboxes, etc.  It needs action listeners and probably many other things.

Reply:
The first comment is about unstructured data. Any class with string fields could arguably be represented as a Map. Hence, rather than representing data as a map, more structure can be added. For example, `map.get("x")`  and `Tuple.getFirst()` are both less structured variants of `Point.getX()`

> And that api doesn't come close to encompassing what the ProAi needs for its debug menu and dialog.

It was not meant to be a comprehensive example. I would not worry about ProAi for just now and focus on creating a clean interface for the new debug and avoid adding to any existing problems (particularly at this early phase were patterns are being set and would be continue further consolidating any anti-patterns that we carry-forward). 

`UiMenuItem` seems to still be very coupled to UI. As a suggestion, here is a slightly more fleshed out example for interfaces that game-core could provide:

```
interface AiDebugMenu {
   Map<PlayerType, AiPlayerDebugOption> getPerPlayerDebugOptions;
   Set<AiDebugOption> getAiDebugOptions;
}

interface AiPlayerDebugOption {    
    /** Used by game-core to set the menu item title */
    String getTitle();

    /** This is a callback into the AI code that should tell it to invoke the renderDelegate to do debug */
    void setEnabled(String playerName, boolean enabled);

    /** This will be called by game-core to provide a render delegate to the AI code so it can access drawing routines */
    void setRenderDelegate(DebugMapRenderer renderDelegate);

    /** This set to true converts us from checkbox to radioButton */
    boolean isMutuallyExclusive() ;

    /** This returning a non-null and/or a non-empty list makes the current object a parent of a checkbox or a radio button list */
    List<AiPlayerDebugOption> getSubCategoryDebugOptions();
}

/** interface for providing an API for AI to draw on the map */
interface DebugMapRenderer {
     void drawOnTerritory(String territoryName, String textValue);
     void colorTerritory(String territoryName, Color color, double transparency);
     void clear();
}

/** Interface for debug option that is generic to AI, EG: 'draw lancaster'
interface AiDebugOption {
    String getOptionTitle();
    void setEnabled(boolean enabled);
    void setRenderDelegate(DebugMapRenderer renderDelegate);
    /** This returning a non-empty list will make the current debug option a parent of a checkbox or a radio button list */
    List<AiDebugOption> getSubCategoryDebugOptions();
}
```

In above, for game-core to render game menus, it would take a collection of debug options and check:
- does this have children? If so, it's the parent of a sub-menu
- for all remaining, group them into two:
   - mutually exclusive (radio button group)
   - not mutually exclusive (check boxes)

Example implementation for  'DrawLancasterDebugOption':
```

@AllArgsConstructor
class DrawLancasterDebugOption {
   private final MapData mapData; // or collection of territories, the base data AI is using
   private boolean enabled;
   private DebugMapRenderer debugMapRenderer;

   void setRenderDelegate(DebugMapRenderer renderDelegate) {
      this.debugMapRenderer = debugMapRenderer;
   }

   boolean setEnabled(boolean enabled) {
        this.enabled = enabled;
        if(enabled) {
             drawDebugToMap();
         }
   }
   
   void onUpdate() {
         if(enabled) {
             drawDebugToMap();
         }
    }

    private void drawDebugToMap() {
           Map<String, Color> territoryShadings = ...;
          territoryShadings.entrySet.forEach(entry -> renderDelegate.shadeTerritory(entry.getKey(), entry.getValue());         
    }
}
```

Hopefully that gives a more useful and concrete example. Please follow up with comments & questions for further discussion.